RFC: Add `no_entry` attribute to omit entry point symbol (#2735)
rust-lang/rfcs: text/0000-no-entry.md

- Feature Name: `no_entry`
- Start Date: 2019-07-23
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

> Add a new top-level attribute `no_entry`, to omit the platform entry point

Reviewer: ```suggestion
Add a new level inner attribute `no_entry`, to omit the platform entry point
```

Reviewer 2: ```suggestion
Add a new crate-level inner attribute `no_entry`, to omit the platform entry point
```

Author: Incorporated (sans "inner"). Thanks!

> symbol, allowing the user to write their own.

Reviewer: ```suggestion
symbol in binary crates, allowing the user to write their own.
```

Author: Done, thanks!

# Motivation
[motivation]: #motivation

For some low-level systems-programming use cases, a user needs to have full
control of a process from the very first instruction, by specifying the
entry-point symbol. Examples include kernels or firmware where the entry-point
symbol may represent initial boot code, applications running under an operating
system but with unusual startup requirements, or programs creating or running
in sandbox environments.

The target toolchain will commonly supply an entry-point symbol (e.g.
`_start`) by default. As a result, attempting to define an entry-point symbol
typically results in a link error. Addressing this currently requires linking
manually with an toolchain-specific argument like `-nostartfiles`.

This proposal introduces a top-level attribute `no_entry`, which will cause the
Rust toolchain to omit the entry-point symbol. Specifying `#![no_entry]` allows
(and requires) the user to supply such a symbol.

(Please note that this is entirely distinct from the Rust-supplied startup hook

> `#[start]`.)

Reviewer 3: Can this section say something about why compiling code as a static library and linking afterwards is insufficient?

Author: @nagisa The later Rationale section discusses that:

> Today, people do this either by linking their binary manually, or by specifying non-portable linker arguments in `.cargo/config`. Introducing the `no_entry` attribute substantially simplifies creating such programs, and avoids target-specific build steps.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Specifying the attribute `#![no_entry]` at the top level of a Rust binary crate
will cause Rust to omit the entry-point symbol when compiling and linking the
binary. When linking, Rust will pass any necessary target-specific arguments to
the toolchain to omit the target's entry-point code, such as `-nostartfiles`.

> Binaries built with this attribute set will typically not link unless the user

Reviewer: Elaborate on typically?

supplies their own entry-point symbol.

Libraries cannot use this attribute; attempting to do so will produce a
compile-time error stating that `no_entry` only applies to binaries.

A binary built with this attribute set could supply the entry-point symbol
itself, or have the entry-point symbol supplied by a library crate or by a
non-Rust library.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

Until stabilized, the `no_entry` attribute will require a feature-gate named
`no_entry`.

> The `no_entry` attribute may be specified as part of a `cfg_attr` declaration,

Reviewer: ```suggestion
The `no_entry` attribute may be the result of expanding a `cfg_attr` attribute
```

> to omit the entry point in some configurations but not others.

Reviewer: ```suggestion
when the `cfg_attr` attribute occurs where `no_entry` is allowed.
```
(also move this paragraph after the one after since you are burying the lede)

Author: Revised and simplified this paragraph, and moved it later as suggested.

Specifying the `no_entry` attribute on anything other than the top-level module

> of a binary crate will produce an error (`warning: crate-level attribute should

Reviewer 3: s/warning: /error: /

Author: Done; thanks!

be in the root module`).

> Declaring an entry-point symbol will typically require a declaration with

Reviewer: Elaborate re. typically.

Reviewer 4: @Centril I understood "typically" here to mean the same thing as it does, for example, when we say that [`main` is typically a function](https://github.com/rust-lang/rust/issues/28179#issuecomment-361320369). The symbol must exist, whether its a function, a static, or "something else" in some future version of Rust is up to the user.

Reviewer: @gnzlbg sure; I'm asking for *textual* changes tho to clarify what it means. :)

> `#[no_mangle] pub extern "C"`.

Reviewer 3: If this attribute removes `_start` glue, then this won’t suffice as `_start` on, I believe, most targets, has atypical calling conventions.

Author: @nagisa At least on Linux, `_start` uses the normal platform calling convention. (Linux uses a separate mechanism, the auxiliary vector, to pass some additional information, but that doesn't affect the calling convention of `_start`.)

On what platforms does `_start` or equivalent require an unusual calling convention?

In any case, writing a `_start` that meets the platform's requirements is the program's responsibility (or Rust's later if we add an `#[entry]` attribute).

Reviewer 5: Linux definitely does not use the standard calling convention for `_start`. On program start, all registers are 0 except for the stack pointer and instruction pointer. The stack contains the following as pointer-sized values, laid out one after the other:
- `argc`
- `argv[..]`
- `NULL`
- `env[..]`
- `NULL`
- the auxiliary vector, terminated by an `AT_NULL` entry

This does not match the x86_64 calling convention at all, or in fact any calling convention that passes arguments in registers. It doesn't even match the x86 calling convention which passes everything on the stack since the top value of the stack is expected to be the return address, which is missing here.

Author: @Amanieu The start function isn't allowed to return, so not having a return address doesn't cause a problem. As for the arguments, people don't think of `_start` as a function that takes those various arguments via a non-standard calling convention; people don't do that in C either. People think of `_start` as a function that takes no arguments, and has to go poke around on the stack itself to obtain the values passed in by the kernel.

In that model, the signature of `_start` on Linux is:

```rust
#[no_mangle] pub extern "C" fn _start() -> !
```

Given the variable argument list, any approach we use to retrieve the arguments will require some iteration within `_start`. (We could potentially use those arguments in-place, such as treating the adjacent `argv` pointers as a single slice, but we'd still need to retrieve them off the stack directly.) Programs that want to replace `_start` will want control over that iteration, rather than having Rust try to translate that into something resembling function parameters.

I can add a note about needing platform-specific code to retrieve any information provided to the entry point function.

Reviewer 6: You're going to have a hard time getting the original stack pointer value to inspect the arguments if `_start` is declared as a Rust function.  You might be able to do it with the equivalent of `__builtin_frame_address` / `llvm.frameaddress`, but I don't think Rust currently exposes that at all, let alone stably.

Anyway, Linux enters the starting point with `rsp` misaligned.  Under the [SysV x86-64 ABI](https://raw.githubusercontent.com/wiki/hjl-tools/x86-psABI/x86-64-psABI-1.0.pdf), on entry to a function, `rsp` is supposed to be a multiple of 16 plus 8.  This is because the caller function is expected to keep `rsp` a multiple of 16, *before* executing the `call` instruction, which subtracts 8 when it pushes the return address onto the stack.  However, upon entry to `_start`, `rsp` "is guaranteed to be 16-byte aligned at process entry".

If you implement `_start` as a Rust (or C) function with the standard calling convention, `rsp` is likely to stay improperly aligned throughout the program, [which can cause mysterious segfaults](https://github.com/rust-lang/rust/issues/59265) if e.g. something tries to use the MOVAPS instruction.

The only valid solutions to this are to implement `_start` in assembly, or to make LLVM support some special "ELF entry point" calling convention.

> To implement `no_entry`, the compiler can have a default linker argument based

Reviewer: ```suggestion
To implement `no_entry`, a Rust compiler can have a default linker argument based
```

on the type of linker in use, such as `-nostartfiles` for GCC or clang/LLD.
Specific targets which require more specific options can provide those options
via the target configuration structures.

Some targets may be able to support invoking the C library without running the

> startup code provided by the C library; other targets may not support this, or

Reviewer: Examples of both include?

Reviewer 5: I don't think any C library supports being invoked without its initialization code. One job of the initialization code is to initialize thread-local storage for the main thread, which is needed for `errno` in particular. This means that calling a C library function that sets `errno` (which is pretty much all of them) will most likely segfault due to TLS not being initialized.

have limitations on such support. Targets that do not support this should
document the target-specific limitations of `no_entry` for their target, may

> require `no_std` in addition to `no_entry`, or may not support `no_entry` at

Reviewer: This also entails that supporting `no_entry` is optional for a Rust compiler in terms of static semantics and that the attribute has primarily implementation defined semantics, right? Would be good to spell that out if so.

Reviewer 4: > This also entails that supporting no_entry is optional for a Rust compiler in terms of static semantics and that the attribute has primarily implementation defined semantics, right?

More than "optional", I understand this document to specify the behavior of `#![no_entry]` as implementation-defined, meaning that the implementation _must_ document what the behavior is.

Often, the language spec constraints the sets of behaviors that can be implemented for implementation-defined behavior, and I believe that it is in the "spirit" of this document to constraint the behaviors that a Rust implementation can provide to:

* compilation error if the implementation does not support `#![no_entry]` for a target 
* compilation error if the implementation does support `#![no_entry]` for a target but using it requires `#![no_std]`
* #![no_entry] does not instantaneously trigger a compilation error, but is handled according to the semantics of the target, which must be documented by the implementation for the particular target according to the text above. This might mean that compilation might fail later, e.g., if some symbol is missing.

That is, a Rust implementation cannot just choose to completely ignore `#![no_entry]` and, e.g., not being even able to parse it. It must explicitly support it, however, what `#![no_entry]` actually then does will depend on the implementation and the target.

all.

On targets that already don't provide an entry point, the compiler will
silently accept `no_entry`. On targets that cannot support `no_entry`, the
compiler will emit a compile-time error.

> 

Reviewer 6: Nitpick: I’d prefer if both cases were an error.  Options that have no effect are confusing and tend to lead to cargo-cult programming.

Reviewer: I disagree -- I believe that if the target provides no entry and you have `#![no_entry]` with no effect then the compiler has fulfilled its contract. I think it makes sense from a declarative specification of behavior (and `#![no_entry]` is a description, not a command).

Author: What @Centril said. I'd like to avoid having to special-case platforms that never have an entry point, when writing code that should never have an entry point. You wrote `no_entry` and the program will have no entry, so the compiler can just say "OK, done, that was easy".

Reviewer: @joshtriplett It would be good to reflect resolving the conversation in the text also with the discussion as material. :)

# Drawbacks
[drawbacks]: #drawbacks

Rust has several existing attributes related to program startup, and this would
require further careful documentation to distinguish them. However, no existing
attribute serves the function that `no_entry` does.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

Today, people do this either by linking their binary manually, or by specifying

> non-portable linker arguments in `.cargo/config`. Introducing the `no_entry`

Reviewer: To ascertain the motivation/rationale of this proposal relative to the complexity cost to the compiler (not users) it would be good to include examples of both to see how much of a hassle existing solutions are. Examples could be included in an appendix to this RFC or inline depending on how much space it takes textually.

> attribute substantially simplifies creating such programs, and avoids

Reviewer: ```suggestion
attribute simplifies creating such programs, and avoids
```
Provide said examples and let them speak for themselves. :)

target-specific build steps.

The name `no_entry` was selected specifically to minimize conflict and

> confusion with other Rust features such as `start`, `no_main`, and `main`, none

Reviewer: To better understand the cost of this proposal, I would like to see an appendix of all the current knobs (stable and unstable, with a note of which is which) regarding startup / entrypoint behavior. Ideally appendix would include a "flow chart" description.

Reviewer 4: > To better understand the cost of this proposal, I would like to see an appendix of all the current knobs (stable and unstable, with a note of which is which) regarding startup / entrypoint behavior

I think it would be reasonable to have all the current knobs documented in a central place somewhere (e.g. the rustc user guide). I think it would be worth it to open an issue in rust-lang/rust about this. Once that happens, this and future RFCs can just reference to those docs and mention the additions.

Reviewer: Sure, that also works.

of which relate to the entry-point symbol.

We could potentially integrate this mechanism with specification of an

> alternate entry point. However, this would make it difficult to flexibly

Reviewer: Elaborate re. what that might look like?

provide the entry point from a library or other mechanism.

> We could make this a rustc command-line option, rather than an attribute.

Reviewer: ```suggestion
We could make this a `rustc` command-line option, rather than an attribute.
```

However, that would separate the configuration required to build a crate from

> the crate itself. We have substantial precedent for including such

Reviewer: I believe this is true but enumerate some examples nonetheless?

configuration directly in the crate as an attribute.

# Prior art
[prior-art]: #prior-art

We currently have the `no_main` attribute to omit a C `main` function (not the
entry point symbol), the `main` attribute to specify a *Rust* `main` function
other than `fn main()`, and the `start` attribute to specify a Rust-specific
portable startup function that runs after some existing Rust startup code.

We do not currently have any mechanisms to address handling of entry-point
symbols.

Existing programming languages rely entirely on the toolchain for this

> functionality. Introducing `no_entry` makes it possible to gain control from

Reviewer 4: Can you provide a minimal example of how C (e.g. using clang) handles this? Do C compilers offer compiler extensions to make this simpler ?

the initial binary entry point without toolchain-specific code, consistent with
other Rust efforts to provide some uniformity across targets and toolchains.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

Can we use common code to implement this for all targets using a given
toolchain family (`LinkerFlavor`), and minimize target-specific implementation
code?

# Future possibilities
[future-possibilities]: #future-possibilities

We should simplify the declaration of an entry-point symbol, such as by

> providing an `#[entry]` attribute. The symbol itself would still have a

Reviewer: If we have `#[entry]` on `unsafe ... fn foobar(...)` do we then also need `#![no_entry]`?

Reviewer 7: FWIW this is an elaboration of https://github.com/rust-lang/rfcs/pull/2735/files/e4ad78dd444961bbc268f4f4a46189999750a5ca#r308998642

Author: @Centril Yes, we do. `entry` could be used in a library, for instance, with a binary crate using `no_entry` and using that library. I'll add a note that it would make sense to have a lint if you use `entry` in a binary crate without using `no_entry`, though.

Reviewer: @joshtriplett Ah I see; so this is akin to "needs / provides" of `#![requires_allocator]` (because every program requires an entry point) and `#[global_allocator]`. -- Please mention this similarity explicitly as well as the use-case for library/binary division in the text also. :)

Author: What do you mean by `requires_allocator`? I don't see any references to that or similar names anywhere on the web.

The RFC already mentions the use case of library/binary division, in the rationale and alternatives section: "However, this would make it difficult to flexibly provide the entry point from a library or other mechanism."

Reviewer: > What do you mean by `requires_allocator`? I don't see any references to that or similar names anywhere on the web.

Ooops; It's called `#![needs_allocator]` in https://github.com/rust-lang/rust/blob/master/src/liballoc/lib.rs#L65 apparently.

> The RFC already mentions the use case of library/binary division, in the rationale and alternatives section: "However, this would make it difficult to flexibly provide the entry point from a library or other mechanism."

Yes, but connecting this to `entry` and `no_entry` is another matter, it would be helpful to clarify that this is what is referred to by that sentence.

> non-portable signature, but having #[entry] would allow using exclusively Rust

Reviewer: ```suggestion
non-portable signature, but having `#[entry]` would allow using exclusively Rust
```

> cfg to handle portability, rather than also needing linker arguments. Rust

Reviewer: ```suggestion
`cfg` to handle portability, rather than also needing linker arguments. Rust
```

could also (optionally) detect and warn about entry points with an unexpected
signature for the target.  This would also make it easier for Rust library

> crates to provide entry points.  (This differs from #[start], which provides a

Reviewer: ```suggestion
crates to provide entry points.  (This differs from `#[start]`, which provides a
```
